File: app/src/main/java/zelpage/razeni/GetTrainAsyncTask.java
package zelpage.razeni;

import java.util.ArrayList;
import java.util.List;

import zelpage.razeni.bo.Couple;
import zelpage.razeni.bo.Train;
import zelpage.razeni.bo.Vehicle;

import android.os.AsyncTask;
import android.text.Html;
import android.view.View;
import android.widget.ArrayAdapter;
import android.widget.EditText;
import android.widget.ListView;
import android.widget.TextView;
import android.widget.Toast;

/**
 * Async task for getting train
 * 
 * @author jindrichbr
 */
public class GetTrainAsyncTask extends AsyncTask<String, Void, List<Train>> {

	private View parent;
	private Train trainResult;
	private TrainFormations api;

	public GetTrainAsyncTask(View p, Train t) {
		this.parent = p;
		this.trainResult = t;

		this.api = new TrainFormations();
	}

	@Override
	protected List<Train> doInBackground(String... trainNumber) {
		if (this.trainResult == null) {
			try {
				List<Train> trains = api.getTrain(trainNumber[0]);
				return trains;

			} catch (Exception ex) {
				return new ArrayList<Train>();
			}
		} else {
			List<Train> r = new ArrayList<Train>();
			r.add(this.trainResult);
			return r;
		}
	}

	@Override
	protected void onPostExecute(List<Train> result) {

		this.parent.findViewById(R.id.progress).setVisibility(View.GONE);

		if (result.size() == 1) {

			EditText traintNumberTB = (EditText) this.parent
					.findViewById(R.id.trainnumtb);
			traintNumberTB.setText("");

			this.trainResult = result.get(0);
			// process response
			TextView trainInfo = (TextView) this.parent
					.findViewById(R.id.label);
			TextView couples = (TextView) this.parent
					.findViewById(R.id.couples);
			TextView deparatures = (TextView) this.parent
					.findViewById(R.id.deparatures);
			this.parent.findViewById(R.id.deparatureslabel).setVisibility(View.VISIBLE);
			TextView notes = (TextView) this.parent.findViewById(R.id.notes);

			if (this.trainResult.getName().compareTo("") == 0) {
				trainInfo.setText(String.format("%s %s",
						this.trainResult.getCategory(),
						this.trainResult.getNumber()));
			} else {
				trainInfo.setText(String.format("%s %s '%s'",
						this.trainResult.getCategory(),
						this.trainResult.getNumber(),
						this.trainResult.getName()));
			}
			trainInfo.invalidate();

			// main couples
			String htmlData = "";
			if (this.trainResult.getCouples().size() == 1) {
				Couple c = this.trainResult.getCouples().get(0);

				for (Vehicle v : c.getVehicles()) {
					htmlData += String.format("<b>%s</b> <i>%s</i>",
							v.getName(), v.getAnnotation());
					htmlData += "<br/>";
				}
			} else {
				for (Couple c : this.trainResult.getCouples()) {

					if (c.getFromStation() != null && c.getToStation() != null) {
						htmlData += String.format(
								"Řazení v úseku %s - %s<br/>",
								c.getFromStation(), c.getToStation());
					}
					for (Vehicle v : c.getVehicles()) {
						htmlData += String.format("<b>%s</b> <i>%s</i>",
								v.getName(), v.getAnnotation());
						htmlData += "<br/>";
					}
					htmlData += "<br/>";
				}
				htmlData = htmlData.substring(0, htmlData.lastIndexOf("<br/>"));
			}
			couples.setText(Html.fromHtml(htmlData));

			String nText = "";
			for (String singleNote : this.trainResult.getNotes()) {
				nText += singleNote + "\n";
			}
			notes.setText(nText);

			nText = "";
			for (String singleNote : this.trainResult.getDepartures()) {
				nText += singleNote + "\n";
			}
			deparatures.setText(nText);
			return;
		} else if (result.isEmpty()) {
			Toast.makeText(this.parent.getContext(), "Žádný vlak nenalezen",
					Toast.LENGTH_LONG).show();
		} else {
			ListView lv = (ListView) this.parent
					.findViewById(R.id.trainsselector);
			lv.setDividerHeight(1);

			ArrayAdapter<Train> data = new ArrayAdapter<Train>(
					this.parent.getContext(), R.layout.listitem, result);
			lv.setAdapter(data);
			lv.setVisibility(View.VISIBLE);
		}

		// hide all
		TextView trainInfo = (TextView) this.parent.findViewById(R.id.label);
		TextView couples = (TextView) this.parent.findViewById(R.id.couples);
		TextView deparatures = (TextView) this.parent
				.findViewById(R.id.deparatures);
		this.parent.findViewById(R.id.deparatureslabel)
				.setVisibility(View.GONE);
		TextView notes = (TextView) this.parent.findViewById(R.id.notes);
		trainInfo.setText("");
		couples.setText("");
		notes.setText("");
		deparatures.setText("");

	}

	@Override
	protected void onPreExecute() {
		this.parent.findViewById(R.id.progress).setVisibility(View.VISIBLE);
		this.parent.findViewById(R.id.trainsselector).setVisibility(View.GONE);
		// hide all
		TextView trainInfo = (TextView) this.parent.findViewById(R.id.label);
		TextView couples = (TextView) this.parent.findViewById(R.id.couples);
		TextView deparatures = (TextView) this.parent
				.findViewById(R.id.deparatures);
		this.parent.findViewById(R.id.deparatureslabel)
				.setVisibility(View.GONE);
		TextView notes = (TextView) this.parent.findViewById(R.id.notes);
		trainInfo.setText("");
		couples.setText("");
		notes.setText("");
		deparatures.setText("");
	}

	public Train getTrainResult() {
		return trainResult;
	}
}

File: app/src/main/java/zelpage/razeni/GetTrainAsyncTask.kt
package zelpage.razeni

import java.util.ArrayList

import android.os.AsyncTask
import android.text.Html
import android.view.View
import android.widget.ArrayAdapter
import android.widget.EditText
import android.widget.ListView
import android.widget.TextView
import android.widget.Toast
import org.jetbrains.anko.AnkoLogger
import org.jetbrains.anko.info

class GetTrainAsyncTask(val parent : View, val t : Train?) : AsyncTask<String, Void, List<Train>>(), AnkoLogger {

	var trainResult : Train? = null
		private set
	private val api : TrainFormations

	init {
		this.trainResult = t

		this.api = TrainFormations()
	}

	override fun doInBackground(vararg trainNumber : String) : List<Train> {
		if (this.trainResult == null) {
			try {
				val trains = api.getTrain(trainNumber[0])
				return trains
			} catch (ex : Exception) {
				info(ex.message, ex)

				return ArrayList()
			}
		} else {
			val r = ArrayList<Train>()
			r.add(this.trainResult!!)
			return r
		}
	}

	override fun onPostExecute(result: List<Train>) {

		this.parent.findViewById(R.id.progress).visibility = View.GONE

		if (result.size == 1) {

			val traintNumberTB = this.parent.findViewById(R.id.trainnumtb) as EditText
			traintNumberTB.setText("")

			this.trainResult = result[0]

			val trainInfo = this.parent.findViewById(R.id.label) as TextView
			val couples = this.parent.findViewById(R.id.couples) as TextView
			val deparatures = this.parent.findViewById(R.id.deparatures) as TextView
			this.parent.findViewById(R.id.deparatureslabel).visibility = View.VISIBLE
			val notes = this.parent.findViewById(R.id.notes) as TextView

			if (this.trainResult!!.name!!.compareTo("") == 0) {
				trainInfo.text = String.format("%s %s",
					this.trainResult!!.category,
					this.trainResult!!.number)
			} else {
				trainInfo.text = String.format("%s %s '%s'",
					this.trainResult!!.category,
					this.trainResult!!.number,
					this.trainResult!!.name)
			}
			trainInfo.invalidate()

			var htmlData = ""
			if (this.trainResult!!.couples!!.size == 1) {
				val c = this.trainResult!!.couples!![0]

				for ((name, annotation) in c.vehicles!!) {
					htmlData += String.format("<b>%s</b> <i>%s</i>",
						name, annotation)
					htmlData += "<br/>"
				}
			} else {
				for ((fromStation, toStation, vehicles) in this.trainResult!!.couples!!) {

					if (fromStation != null && toStation != null) {
						htmlData += String.format(
							"Řazení v úseku %s - %s<br/>",
							fromStation, toStation)
					}
					for ((name, annotation) in vehicles!!) {
						htmlData += String.format("<b>%s</b> <i>%s</i>",
							name, annotation)
						htmlData += "<br/>"
					}
					htmlData += "<br/>"
				}
				htmlData = htmlData.substring(0, htmlData.lastIndexOf("<br/>"))
			}
			couples.text = Html.fromHtml(htmlData)

			var nText = ""
			for (singleNote in this.trainResult!!.notes!!) {
				nText += singleNote + "\n"
			}
			notes.text = nText

			nText = ""
			for (singleNote in this.trainResult!!.departures!!) {
				nText += singleNote + "\n"
			}
			deparatures.text = nText
			return
		} else if (result.isEmpty()) {
			Toast.makeText(this.parent.context, "Žádný vlak nenalezen",
				Toast.LENGTH_LONG).show()
		} else {
			val lv = this.parent.findViewById(R.id.trainsselector) as ListView
			lv.dividerHeight = 1

			val data = ArrayAdapter(
				this.parent.context, R.layout.listitem, result)
			lv.adapter = data
			lv.visibility = View.VISIBLE
		}

		val trainInfo = this.parent.findViewById(R.id.label) as TextView
		val couples = this.parent.findViewById(R.id.couples) as TextView
		val deparatures = this.parent.findViewById(R.id.deparatures) as TextView
		this.parent.findViewById(R.id.deparatureslabel).visibility = View.GONE
		val notes = this.parent.findViewById(R.id.notes) as TextView
		trainInfo.text = ""
		couples.text = ""
		notes.text = ""
		deparatures.text = ""
	}

	override fun onPreExecute() {
		this.parent.findViewById(R.id.progress).visibility = View.VISIBLE
		this.parent.findViewById(R.id.trainsselector).visibility = View.GONE
		// hide all
		val trainInfo = this.parent.findViewById(R.id.label) as TextView
		val couples = this.parent.findViewById(R.id.couples) as TextView
		val deparatures = this.parent.findViewById(R.id.deparatures) as TextView
		this.parent.findViewById(R.id.deparatureslabel).visibility = View.GONE
		val notes = this.parent.findViewById(R.id.notes) as TextView
		trainInfo.text = ""
		couples.text = ""
		notes.text = ""
		deparatures.text = ""
	}
}
